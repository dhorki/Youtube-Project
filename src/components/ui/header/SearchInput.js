import React from 'react';
import styled from 'styled-components';
import { BsSearch } from 'react-icons/bs';
import { useForm } from '../../../hooks/useForm';
import { MdClear } from 'react-icons/md';
import { useTheme } from '../../../hooks/useTheme';

const StyledSearchBox = styled.form`
  background-color: ${({ theme }) => theme.inputBackground};
  border-radius: 2px;
  border: 1px solid ${({ theme }) => theme.inputBorder};
  display: flex;
  flex-direction: row;
  margin: auto;
  height: 30px;

  > input {
    color: ${({ theme }) => theme.primary};
    font-family: 'Roboto', 'Arial', sans-serif;
    border: none;
    border-radius: 2px;
    background-color: transparent;
    padding: 0.5rem;
    padding-right: 0.2rem;
    width: ${({ q }) => (q.length === 0 ? 380 : 350)}px;

    &:focus {
      outline: none;
    }
  }

  > .clear-button {
    ${({ q }) => q.length === 0 && 'display: none;'}
    width: 30px;
    color: ${({ theme }) => theme.primary};
    background-color: transparent;
    border: none;
    cursor: pointer;
    height: 100%;
    margin-top: auto;
    margin-bottom: auto;
  }

  > .search-button {
    color: ${({ theme }) => theme.primary};
    background-color: ${({ theme }) => theme.searchButtonBackground};
    border: none;
    border-left: 1px solid ${({ theme }) => theme.inputBorder};
    cursor: pointer;
    justify-content: center;
    padding: 0;
    margin: auto;
    width: 50px;
    height: 100%;
  }
`;

export const SearchInput = () => {
  const { getTheme } = useTheme();
  const theme = getTheme();
  // TODO: initialize the form with the initial values
  const [{ q }, handleInputChange] = useForm({ q: '' });

  const handleSearch = (e) => {
    e.preventDefault();
    console.log('Search', q);
  };

  const handleClear = (e) => {
    e.preventDefault();
    handleInputChange({ target: { name: 'q', value: '' } });
  };

  return (
    <StyledSearchBox theme={theme} q={q}>
      <input
        type="text"
        autoComplete="off"
        placeholder="Search"
        value={q}
        onChange={handleInputChange}
        name="q"
        theme={theme}
      />
      <button className="clear-button" type="button" onClick={handleClear}>
        <MdClear />
      </button>

      <button className="search-button" type="submit" onClick={handleSearch}>
        <BsSearch />
      </button>
    </StyledSearchBox>
  );
};
