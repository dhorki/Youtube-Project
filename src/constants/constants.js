export const styles = {
  colors: {
    light: {
      appIcon: '#ff0200',
      background: '#F9F9F9',
      barBackground: '#ffffff',
      primary: '#030303',
      secondary: '#606060',
      linkActive: '#ff0200',
      linkHoverBackground: '#ececec',
      inputBackground: '#ffffff',
      inputBorder: '#cccccc',
      searchButtonBackground: '#f8f8f8',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
    dark: {
      appIcon: '#ff0200',
      background: '#181818',
      barBackground: '#212121',
      primary: '#ffffff',
      secondary: '#aaaaaa',
      linkActive: '#ff0200',
      linkHoverBackground: '#303030',
      inputBackground: '#121212',
      inputBorder: '#303030',
      searchButtonBackground: '#323232',
      // google button
      white: '#fff',
      googleBlue: '#4285f4',
      buttonActiveBlue: '#1669f2',
    },
  },
  props: {
    transitionTime: '0.4s',
    sidebarWidth: '150px',
  },
};

export const screenSizes = {
  xsmall: '675px',
  small: '830px',
  mid: '1200px',
};

export const mainViewTypes = {
  searchGallery: 'searchGallery',
  favoritesGallery: 'favoritesGallery',
  videoDetails: 'videoDetails',
};

export const environmentActions = {
  setTheme: 'setTheme',
  setSidebarShow: 'setSidebarShow',
  setSearchQuery: 'setSearchQuery',
  setModalShow: 'setModalShow',
  setUser: 'setUser',
};

export const modalTypes = {
  login: 'login',
  register: 'register',
  resetPassword: 'reset-password',
  none: 'none',
};

export const resetSearchQueryAction = {
  type: environmentActions.setSearchQuery,
  payload: {
    q: '',
  },
};

export const hideModalAction = {
  type: environmentActions.setModalShow,
  payload: {
    modalShow: modalTypes.none,
  },
};
