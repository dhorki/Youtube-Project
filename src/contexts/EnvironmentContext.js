import { createContext } from 'react';

export const EnvironmentContext = createContext();
