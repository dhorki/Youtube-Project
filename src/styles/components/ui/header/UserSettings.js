import styled from 'styled-components';

export const StyledUserSettings = styled.div`
  display: flex;
  flex-direction: row;
  padding-right: 30px;
  justify-content: flex-end;

  > .username-tag {
    color: ${({ theme }) => theme.primary};
    font-size: 15px;
    font-weight: 400;
    margin: auto;
    margin-right: 10px;
  }

  > .user-badge {
    background-color: red;
    border: 1px solid ${({ theme }) => theme.inputBorder};
    border-radius: 50%;
    height: 30px;
    width: 30px;
    margin: auto;
    cursor: pointer;
  }
`;
