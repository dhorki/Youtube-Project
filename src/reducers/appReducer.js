import { environmentActions } from '../constants/constants';

export const appReducer = (state, { type, payload }) => {
  switch (type) {
    case environmentActions.setTheme:
      return { ...state, theme: payload.theme };
    case environmentActions.setSidebarShow:
      return { ...state, sidebarShow: payload.sidebarShow };
    case environmentActions.setSearchQuery:
      return { ...state, q: payload.q };
    default:
      return state;
  }
};
