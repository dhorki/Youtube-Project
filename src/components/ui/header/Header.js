import React, { useContext } from 'react';
import { SearchInput } from './SearchInput';
import { UserSettings } from './UserSettings';
import { IoLogoYoutube } from 'react-icons/io5';
import { GiHamburgerMenu } from 'react-icons/gi';
import { BsFillArrowLeftSquareFill } from 'react-icons/bs';
import { Link } from 'react-router-dom';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';
import { StyledHeader } from '../../../styles/components/ui/header/Header';
import { environmentActions } from '../../../constants/constants';
import { resetSearch } from '../../../helpers/resetSearch';

export const Header = () => {
  const { environment, dispatchEnv } = useContext(EnvironmentContext);
  const { theme, sidebarShow } = environment;

  const sidebarShowHandler = () => {
    const action = {
      type: environmentActions.setSidebarShow,
      payload: {
        sidebarShow: !sidebarShow,
      },
    };

    dispatchEnv(action);
  };

  return (
    <StyledHeader theme={theme}>
      <button id="sidebarToggler" onClick={sidebarShowHandler}>
        {sidebarShow ? <BsFillArrowLeftSquareFill /> : <GiHamburgerMenu />}
      </button>
      <Link className="app-link" to="/" onClick={() => resetSearch(dispatchEnv)}>
        <div className="app-icon">
          <IoLogoYoutube />
        </div>
        <h1>Llutuve</h1>
      </Link>
      <SearchInput />
      <UserSettings />
    </StyledHeader>
  );
};
