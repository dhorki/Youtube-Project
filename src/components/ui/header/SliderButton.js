import React, { useContext, useState } from 'react';
import PropTypes from 'prop-types';
import { StyledSliderButton } from '../../../styles/components/ui/header/SliderButton';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';

export const SliderButton = ({ label, initState, callback }) => {
  const { environment } = useContext(EnvironmentContext);
  const { theme } = environment;

  const [hover, setHover] = useState(false);
  const [checked, setChecked] = useState(initState);

  const handleToggle = () => {
    setChecked(!checked);
    if (callback) {
      callback(!checked);
    }
  };

  return (
    <StyledSliderButton
      onMouseOver={() => {
        setHover(true);
      }}
      onMouseOut={() => {
        setHover(false);
      }}
      onClick={handleToggle}
      hover={hover}
      checked={checked}
      theme={theme}
    >
      <div className="slider-background">
        <div className="slider" />
      </div>
      {label && <div className="slider-label">{label}</div>}
    </StyledSliderButton>
  );
};

SliderButton.propTypes = {
  label: PropTypes.string,
  initState: PropTypes.bool.isRequired,
  callback: PropTypes.func,
};
