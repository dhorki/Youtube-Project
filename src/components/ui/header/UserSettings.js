import React, { useContext, useState } from 'react';
import { RiLoginBoxFill, RiLogoutBoxFill } from 'react-icons/ri';
import { SliderButton } from './SliderButton';
import { EnvironmentContext } from '../../../contexts/EnvironmentContext';
import { environmentActions, modalTypes, styles } from '../../../constants/constants';
import { StyledUserSettings } from '../../../styles/components/ui/header/UserSettings';
import { logout } from '../../../firebase';

export const UserSettings = () => {
  const { environment, dispatchEnv } = useContext(EnvironmentContext);
  const { theme, user } = environment;

  const [submenuShow, setsubmenuShow] = useState(false);

  const handleThemeChange = (checked) => {
    const action = {
      type: environmentActions.setTheme,
      payload: {
        theme: checked ? styles.colors.dark : styles.colors.light,
      },
    };

    dispatchEnv(action);
  };

  const handleLoginLogout = (toLogin) => {
    const action = {
      type: environmentActions.setModalShow,
      payload: {
        // if not toLogin, then logout, logout does not need modalShow
        modalShow: toLogin ? modalTypes.login : modalTypes.none,
      },
    };

    if (toLogin) {
      console.log('login');
    } else {
      const action = {
        type: environmentActions.setUser,
        payload: {
          user: null,
        },
      };

      logout();
      dispatchEnv(action);
    }

    setsubmenuShow(false);

    dispatchEnv(action);
  };

  const logged = false;

  return (
    <StyledUserSettings theme={theme} submenuShow={submenuShow}>
      <SliderButton
        label="Dark Mode"
        initState={theme === styles.colors.dark}
        callback={handleThemeChange}
      />
      {user && <div className="username-tag">{user.name}</div>}
      <img
        className="user-badge"
        onClick={() => setsubmenuShow(!submenuShow)}
        alt={user?.name}
        src={
          user?.photo
            ? user.photo
            : 'https://cdn.pixabay.com/photo/2016/11/08/15/21/user-1808597_1280.png'
        }
      />
      <div className="submenu-container">
        <div className="submenu-item" onClick={() => handleLoginLogout(null === user)}>
          <div className="submenu-icon">
            {logged ? <RiLogoutBoxFill /> : <RiLoginBoxFill />}
          </div>
          <div className="submenu-label">{null === user ? 'Login' : 'Logout'}</div>
        </div>
      </div>
    </StyledUserSettings>
  );
};
