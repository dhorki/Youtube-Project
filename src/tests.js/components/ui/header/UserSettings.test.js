import React, { useReducer } from 'react';
import { Router } from 'react-router-dom';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { createBrowserHistory } from 'history';
import { environmentActions, styles } from '../../../../constants/constants';
import { EnvironmentContext } from '../../../../contexts/EnvironmentContext';
import { UserSettings } from '../../../../components/ui/header/UserSettings';
import { renderHook, act } from '@testing-library/react-hooks';
import { appReducer } from '../../../../reducers/appReducer';

describe('UserSettings', () => {
  const environment = {
    theme: styles.colors.light,
    sidebarShow: false,
    q: '',
  };

  it('should render correctly', () => {
    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment }}>
        <Router history={createBrowserHistory()}>
          <UserSettings />
        </Router>
      </EnvironmentContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle change theme to dark', () => {
    const { result } = renderHook(() => useReducer(appReducer, environment));
    const [prevEnv] = result.current;
    const dispatchEnv = jest.fn();

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment: prevEnv, dispatchEnv }}>
        <UserSettings />
      </EnvironmentContext.Provider>
    );

    act(() => {
      wrapper.find('button').simulate('click');
    });

    const action = {
      type: environmentActions.setTheme,
      payload: { theme: styles.colors.dark },
    };

    expect(dispatchEnv).toHaveBeenCalled();
    expect(dispatchEnv).toHaveBeenCalledWith(action);
  });

  it('should handle change theme to light', () => {
    environment.theme = styles.colors.dark;
    const { result } = renderHook(() => useReducer(appReducer, environment));
    const [prevEnv] = result.current;
    const dispatchEnv = jest.fn();

    const wrapper = mount(
      <EnvironmentContext.Provider value={{ environment: prevEnv, dispatchEnv }}>
        <UserSettings />
      </EnvironmentContext.Provider>
    );

    act(() => {
      wrapper.find('button').simulate('click');
    });

    const action = {
      type: environmentActions.setTheme,
      payload: { theme: styles.colors.light },
    };

    expect(dispatchEnv).toHaveBeenCalled();
    expect(dispatchEnv).toHaveBeenCalledWith(action);
  });
});
