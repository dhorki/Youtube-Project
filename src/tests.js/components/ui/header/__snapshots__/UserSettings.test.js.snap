// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSettings should render correctly 1`] = `
<div
  className="sc-gsnTZi kemrEt"
>
  <button
    checked={false}
    className="sc-bczRLJ dqlSGE"
    onClick={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="slider-background"
    >
      <div
        className="slider"
      />
    </div>
    <div
      className="slider-label"
    >
      Dark Mode
    </div>
  </button>
  <div
    className="username-tag"
  >
    Victor Infante
  </div>
  <img
    alt="Victor Infante"
    className="user-badge"
    src="https://cdn.pixabay.com/photo/2016/11/08/15/21/user-1808597_1280.png"
  />
</div>
`;
